packages/plugin-hardhat/src/utils/deploy.ts

```diff
@@ -25,16 +26,27 @@ export async function deploy(
 
 async function ethersDeploy(factory: ContractFactory, ...args: unknown[]) {
   const contractInstance = await factory.deploy(...args);
-  const { deployTransaction } = contractInstance;
-
-  const address: string = getContractAddress({
-    from: await factory.signer.getAddress(),
-    nonce: deployTransaction.nonce,
-  });
-  if (address !== contractInstance.address) {
-    debug(
-      `overriding contract address from ${contractInstance.address} to ${address} for nonce ${deployTransaction.nonce}`,
-    );
+  const deployTransaction = contractInstance.deploymentTransaction();
+
+  if (deployTransaction === null) {
+    throw new Error('Broken invariant: deploymentTransaction is null');
+  }
+
+  const signer = getSigner(factory.runner);
+  const contractAddress = await contractInstance.getAddress();
+
+  let address = contractAddress;
+
+  if (signer !== undefined) {
+    const from = await signer.getAddress();
+
+    address = getCreateAddress({
+      from: from,
+      nonce: deployTransaction.nonce,
+    });
```

Comment:
Why do we need this? I don't remember. If you do, can we add a comment?

Reply:
See https://github.com/OpenZeppelin/openzeppelin-upgrades/issues/767#issuecomment-1496260928